tonic/src/transport/service/timeout.rs

```diff
@@ -0,0 +1,251 @@
+use http::Request;
+use http::{HeaderMap, HeaderValue};
+use pin_project::pin_project;
+use std::{fmt, pin::Pin};
+use std::{
+    future::Future,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::Sleep;
+use tower_service::Service;
+
+#[derive(Debug, Clone)]
+pub(crate) struct Timeout<S> {
+    inner: S,
+    server_timeout: Option<Duration>,
+}
+
+impl<S> Timeout<S> {
+    pub(crate) fn new(inner: S, server_timeout: Option<Duration>) -> Self {
+        Self {
+            inner,
+            server_timeout,
+        }
+    }
+}
+
+impl<S, ReqBody> Service<Request<ReqBody>> for Timeout<S>
+where
+    S: Service<Request<ReqBody>>,
+    S::Error: Into<crate::Error>,
+{
+    type Response = S::Response;
+    type Error = crate::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let client_timeout = try_parse_grpc_timeout(req.headers()).unwrap_or_else(|e| {
+            tracing::trace!("Error parsing `grpc-timeout` header {:?}", e);
```

Comment:
```suggestion
            tracing::trace!("Error parsing `grpc-timeout` header {}", e);
```

Probably want to use `Display` here instead of `Debug`?

Reply:
`HeaderValue` doesn't implement `Display` actually. From skimming the code it seem `Debug` does the appropriate thing for this case.